pilot/pkg/model/context.go

```diff
@@ -208,10 +210,10 @@ var DefaultXdsLogDetails = XdsLogDetails{}
 // Note: any errors returned will completely close the XDS stream. Use with caution; typically and empty
 // or no response is preferred.
 type XdsResourceGenerator interface {
-	Generate(proxy *Proxy, push *PushContext, w *WatchedResource, updates *PushRequest) (Resources, XdsLogDetails, error)
-
-	// GenerateDeltas returns the changed and removed resources, along with whether or not delta was actually used.
-	GenerateDeltas(proxy *Proxy, push *PushContext, updates *PushRequest, w *WatchedResource) (Resources, []string, XdsLogDetails, bool, error)
```

Comment:
I donot see much benefit merging the two functions

Reply:
It bring so many updates

Reply:
The benefit and the main reason I merged is to not force every generator to implement this. Another option I can think of is introduce a NoOpDeltaGenerator struct that just throws "Not implement" but that does not work as delta.go blindly  calls GenerateDeltas if we have that method but it expects some response. 

Is there a way to avoid that boiler plate code for every generator?

Reply:
But actually, you need to change all the existing empty implementations

Reply:
Compared to a bool value to indicate delta, I am more for the separate function

Reply:
This bool value is not a permanent thing. Since we are doing incrementally we needed this. After all is done, we can remove this and mark a generator as "Delta" generator so that we just process "removed resources". Even for CDS today it is needed.

Reply:
IMO it is permanent. Even years from now, there will still be cases where the generators need to just rebuild everything (ex - mesh config changes)

Reply:
But in that case can not we rely on removedResources?

Reply:
I do not think so. Say a mesh config updates. In CDS, we would enter the current logic for buildOutboundClusters (iterate over all services/DR and make clusters). At this point we have a list of clusters; removedResources is empty.

However, in reality, we may have removed some clusters! Since delta=true and removedResources=empty, we do not remove any. This is wrong.

Instead, we need to compute the removedResources from `watchedResources.Names - newClusters.Names`

Reply:
Basically _someone_ needs to compute which clusters are removed. This means either the generators are smart enough to perfectly handle 100% of of possible pushes (never going to happen IMO), or we have a fallback logic of `watchedResources.Names - newClusters.Names` in delta.go

Reply:
Let me go through in detail.  But if it is permanent flag it is  even needed GenerateDeltas. So we can not avoid the boolean by having two separate functions @hzxuzhonghu

Reply:
however, we can avoid having `RemovedResources` in `Generate` if we have a separate API. IMO it's cleaner to keep the existing `Generate` as it is API and not have `RemovedResources` there, delta seems like a separate thing altogether.

Reply:
Ok. If every one things that is better way - I can leave it like that. I have no strong preference. Let me carve out the general refactors in to a separate PR and close this one.